edge-hub/src/Microsoft.Azure.Devices.Routing.Core/Router.cs

```diff
@@ -181,11 +187,24 @@ protected virtual void Dispose(bool disposing)
             }
         }
 
-        static ISet<Endpoint> GetEndpoints(RouterConfig config)
+        static IList<uint> GetPrioritiesForEndpoint(Endpoint endpoint, IEnumerable<Route> routes)
+        {
+            return routes
+                .Where(r => r.Endpoint == endpoint)
+                .Select(r => r.Priority).ToList();
+        }
+
+        static ISet<(Endpoint, IList<uint>)> GetEndpointsWithPriority(RouterConfig config)
         {
-            var endpoints = new HashSet<Endpoint>(config.Routes.Select(r => r.Endpoint));
-            config.Fallback.ForEach(f => endpoints.Add(f.Endpoint));
-            return endpoints;
+            var endpoints = new HashSet<(Endpoint, IList<uint>)>(config.Endpoints.Select(e => (e, GetPrioritiesForEndpoint(e, config.Routes))));
+            config.Fallback.ForEach(f => endpoints.Add((f.Endpoint, new List<uint>() { RouteFactory.DefaultPriority })));
+
+            // De-dupe for the case where the fallback route uses the same
+            // endpoint as a normal route
+            return endpoints
+                .GroupBy(e => e.Item1)
+                .Select(dupe => dupe.First())
```

Comment:
should this be a merge of all the dupes, instead of just taking the first one and discarding the rest?

Reply:
just realized that this is prob better served as a dictionary, then we don't have the deduping concerns

---
In reply to: [381575308](https://github.com/Azure/iotedge/pull/2525#discussion_r381575308) [](ancestors = 381575308)